### PR TITLE
Match OSX behaviour of fullscreen: false on Windows and Linux

### PR DESCRIPTION
**Pull request by brenca**
This won't cancel the `enter-html-full-screen` event, so a youtube video will still go fullscreen, but will be contained inside the `BrowserWindow`.

### CODE REVIEW

**Comment by sindresorhus** [comment]
Why not just make the existing `enter-full-screen` event cancellable?

**Comment by brenca** [comment]
Ofc that was my initial idea, but to do that, we'd need to move some code around that triggers that event on osx (it fires after the windows is already fullscreen) and I didn't want to break anyone's code that relies on that behaviour.

**Comment by zcbenz** [comment]
This event is a bit meaningless to me, it can not prevent window from entering fullscreen when users click the fullscreen button on OS X and Linux, what it does is to add a hook in `setFullScreen`, which can be easily done on users' side.

**Comment by brenca** [comment]
The problem I tried to solve with this is that we're using offscreen rendering with a hidden window, and if - for example - we double click on a youtube video, it wants to go fullscreen, that makes the window itself fullscreen which also shows the initally hidden window (it also messes with the buffer's size since the window is resized in the process), and even if I handle the fullscreen event and set it back to it's original state, it flashes for a moment. Can I solve this somehow other than with a cancellable event? I might have missed something.

**Comment by zcbenz** [comment]
From next version you will be able to deny the request of fullscreen:
https://github.com/atom/electron/blob/master/docs/api/session.md#sessetpermissionrequesthandlerhandler

**Comment by brenca** [comment]
I've tried this method with youtube - it has an effect like what I've seen when I tried to make the fullscreen event cancellable - if you click on the fullscreen button nothing happens until you pause the video, then it seems to think that it's actually fullscreen, and the controls and video size all go buggy. With the proposed event, this does not happen anymore. I might look into this and try to make the permissionhandler method work like this, if you think it can be done.

**Comment by zcbenz** [comment]
The problem with the event you proposed is, it doesn't do what it claims to do, which makes it only meaningful for your own use.

We can probably tune the permission handler to make the page think it is in fullscreen mode but actually not.

**Comment by deepak1556** [comment]
@brenca thanks for the test case, Unfortunately html5 fullscreen api doesnt provide a good way to deny permission. The api expects a fullscreenerror event which is triggered under a few scenarios if the device doenst support or if the target element doesnt have the allowfullscreen attribute or if api is called without user gesture. The enterfullscreen call happens async after all these check, related [code](https://code.google.com/p/chromium/codesearch#chromium/src/third_party/WebKit/Source/core/dom/Fullscreen.cpp&cl=GROK&gsn=RequestType&rcl=1455596815&l=200). So with permission handler Electron can help deny the fullscreen animation but the api users will be under the assumption fullscreenchange has taken place as there isnt an error emitted. Any idea to fix this without patching chromium ?

**Comment by xywz** [comment]
@brenca what about `fullscreen: false` when creating the window?

**Comment by brenca** [comment]
@xywz `fullscreen: false` only disables the maximize button as far as I know, it won't force the window to stay not-fullscreen

**Comment by brenca** [comment]
@deepak1556 I looked into it a bit, and I don't think we can do this kind of functionality without some chromium hacking. Also, my ultimate goal with this was not denying the request entirely, but to make the HTML side of this happen while still keeping the native window in it's original state. I'm using frame subscription for offscreen rendering, and at the current state of things, I can either deny everything fullscreen related by denying the permission request (and take this kinda buggy behaviour), or let the fullscreen thing happen, but then my hidden window will be shown and resized, which means a window that should be hidden is shown, and the buffer that I get suddenly grows, which causes problems for me, and I could not find any proper solution to this problem on my side (even if I instantly exit the fullscreen state and hide the window some bad frames will get through, and the window will show itself for a split second in the background). I'm open for suggestions, if anyone has any.

**Comment by xywz** [comment]
@brenca `fullscreen: false` does force the window to stay not-fullscreen.

**Comment by brenca** [comment]
@xywz Well, I've tried that again, and it doesn't even disable the maximize button for me (Win10), it basicly has no effect on the window. `maximizable: false` does disable the button, but I can still use a youtube video to make the window fullscreen.
Also, according to the documentation, `fullscreen: false` does the following: "When explicitly set to false the fullscreen button will be hidden or disabled on OS X, or the maximize button will be disabled on Windows." There's nothing here about forcing the window to stay not-fullscreen.

@zcbenz would it be acceptable if I instead of this event I made `fullscreen: false` do the same thing as cancelling the event does?

**Comment by xywz** [comment]
@brenca it does force the window to stay not-fullscreen on OS X. It makes sense to match that behavior on Windows.

**Comment by zcbenz** [comment]
> @zcbenz would it be acceptable if I instead of this event I made fullscreen: false do the same thing as cancelling the event does?

It seems that `fullscreen: false` already prevents `setFullScreen` from working?

**Comment by brenca** [comment]
I've tried this code on Windows 10

```javascript
mainWindow = new BrowserWindow({
  width: 800,
  height: 600,
  fullscreen: false
});

setTimeout(function() {
  mainWindow.setFullScreen(true)
}, 1000)
```

The window goes fullscreen after setFullScreen is executed. On Mac, `fullscreen: false` seems to work exactly how I want it to, thanks @xywz for pointing that out (my colleague tested it on his Mac too). I'll try to match that behaviour on Windows and update this PR with that code (and title and description).

File: atom/browser/native_window_views.cc [hunk 1]
```diff
@@ -490,11 +496,15 @@ bool NativeWindowViews::IsMaximizable() {
 #endif
 }
 
-void NativeWindowViews::SetFullScreenable(bool maximizable) {
+void NativeWindowViews::SetFullScreenable(bool fullscreenable) {
+  #if defined(OS_WIN)
+    FlipWindowStyle(GetAcceleratedWidget(), false, WS_MAXIMIZEBOX);
```
Comment (zcbenz) [inline atom/browser/native_window_views.cc]: I don't think we need to change the style here, this disables the maximize box, which is not related to fullscreen.

Comment (brenca) [inline atom/browser/native_window_views.cc]: According to the `BrowserWindow` [docs](https://github.com/atom/electron/blob/master/docs/api/browser-window.md#new-browserwindowoptions), `fullscreen: false` should disable the maximize button on Windows, that's why I included this. Shall I remove it and correct the docs?

Comment (zcbenz) [inline atom/browser/native_window_views.cc]: The maximize/zoom button thing is only limited to OS X, on OS X the zoom button is used as fullscreen button by default.

Comment (brenca) [inline atom/browser/native_window_views.cc]: I'll correct the docs and the code then!

**Comment by zcbenz** [comment]
:+1:

**Pull request merged**